streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1761,19 +1761,13 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
   final def interruptWhen[E1 >: E](p: Promise[E1, _]): ZStream[R, E1, O] =
     ZStream {
       for {
-        as   <- self.process
-        done <- Ref.make(false).toManaged_
-        pull = done.get flatMap {
-                 if (_) Pull.end
-                 else
-                   as.raceFirst(
-                     p.await
-                       .mapError(Some(_))
-                       .foldCauseM(
-                         c => done.set(true) *> ZIO.halt(c),
-                         _ => done.set(true) *> Pull.end
-                       )
-                   )
+        as    <- self.process
+        done  <- Ref.makeManaged(false)
+        asPull = p.await.asSomeError *> done.set(true) *> Pull.end
```

Comment:
Interesting. I compared with my [suggestion here](https://github.com/zio/zio/pull/2442/files#r355213169) and that one got #4119 right, but would also experience #4118

Reply:
@svroonland are you sure? it checks on every pull whether the promise is done and there's a unit test for that case.

Reply:
@luis3m Ah yes on closer look, you are right. So they would appear to be equivalent then..? (just for learning purposes, you can disregard this comment thread when it comes to this PR)

Reply:
@svroonland no worries. Yes, they are.